### PR TITLE
Add response attribute extraction

### PR DESCRIPTION
**Pull request by ybasket**
Supports extracting span attributes depending on the sttp response, for example to have HTTP status codes in traces.

### CODE REVIEW

**Comment by catostrophe** [comment]
@ybasket Thanks for the PR!

I reviewed it and realized we had inconsistency between the `sttp` and `tapir` modules wrt adding req/resp attributes to the traces, just because it was done by different people at different times.

My proposal is to change the code a bit:
```scala
val ctxHeaders = headersGetter.get(childCtx)
val req = request.headers(SttpHeaders.converter.to(ctxHeaders).headers: _*)

val isSampled = childSpan.context.traceFlags.sampled == SampleDecision.Include

// only extract uri attributes if the span is sampled as the host parsing is quite expensive
val reqUriAttributes = if (isSampled) SttpRequest.toAttributes(request) else Map.empty

for {
  _ <- childSpan.putAll(
    SttpHeaders.requestFields(Headers(req.headers), dropHeadersWhen) ++ reqUriAttributes: _*
  )
  resp <- lower(ctxBackend.send(req))
  _ <- childSpan.setStatus(SttpStatusMapping.statusToSpanStatus(resp.statusText, resp.code))
  _ <- childSpan.putAll(SttpHeaders.responseFields(Headers(resp.headers), dropHeadersWhen): _*)
} yield resp
```

What we'll have in this case:
- ctx attributes go to both headers and span attributes
- all req and resp headers go to span attributes, with filtering if needed
- we use the same methods as in the `tapir` module


cc @janstenpickle for discussion

**Comment by catostrophe** [comment]
Oh, you needed some extra info apart from headers. Maybe we can combine both enhancements.

**Comment by ybasket** [comment]
> Oh, you needed some extra info apart from headers. Maybe we can combine both enhancements.

Just pushed a commit that combines both along your snippet. I only wonder whether the `putAll(SttpHeaders.responseFields...)` should also be wrapped in a `whenA` to have full lazy attribute handling. WDYT?

File: modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendSyntax.scala [hunk 1]
```diff
@@ -3,30 +3,44 @@ package io.janstenpickle.trace4cats.sttp.client3
...
 
 trait SttpBackendSyntax {
 
   implicit class TracedSttpBackendSyntax[F[_], +P](backend: SttpBackend[F, P]) {
     def liftTrace[G[_]](
       toHeaders: ToHeaders = ToHeaders.standard,
-      spanNamer: SttpSpanNamer = SttpSpanNamer.methodWithPath
+      spanNamer: SttpSpanNamer = SttpSpanNamer.methodWithPath,
+      dropHeadersWhen: String => Boolean = HeaderNames.isSensitive,
+      attributesFromResponse: Getter[Response[Unit], Map[String, AttributeValue]] = Getter(_ => Map.empty)
```
Comment (catostrophe) [inline modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendSyntax.scala:18]:
```suggestion
      responseAttributesGetter: Getter[Response[_], Map[String, AttributeValue]] = Getter(_ => Map.empty)
```

**Comment by catostrophe** [comment]
> should also be wrapped in a `whenA` to have full lazy attribute handling

IMO, there's one case that is worth optimizing -- it's parsing the uri host.

> but maybe still worth adding as an attribute by default

I would prefer default attributes. If so, do we still need an extra getter?

File: modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendSyntax.scala [hunk 2]
```diff
@@ -3,30 +3,44 @@ package io.janstenpickle.trace4cats.sttp.client3
...
+        attributesFromResponse
       )
 
     def liftTraceContext[G[_], Ctx](
       spanLens: Lens[Ctx, Span[F]],
       headersGetter: Getter[Ctx, TraceHeaders],
-      spanNamer: SttpSpanNamer = SttpSpanNamer.methodWithPath
+      spanNamer: SttpSpanNamer = SttpSpanNamer.methodWithPath,
+      dropHeadersWhen: String => Boolean = HeaderNames.isSensitive,
+      attributesFromResponse: Getter[Response[Unit], Map[String, AttributeValue]] = Getter(_ => Map.empty)
```
Comment (catostrophe) [inline modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendSyntax.scala:34]:
```suggestion
      responseAttributesGetter: Getter[Response[_], Map[String, AttributeValue]] = Getter(_ => Map.empty)
```

File: modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala [hunk 3]
```diff
@@ -1,22 +1,26 @@
...
 import sttp.monad.{MonadError => SttpMonadError}
 
 class SttpBackendTracer[F[_], G[_], +P, Ctx](
   backend: SttpBackend[F, P],
   spanLens: Lens[Ctx, Span[F]],
   headersGetter: Getter[Ctx, TraceHeaders],
-  spanNamer: SttpSpanNamer
+  spanNamer: SttpSpanNamer,
+  dropHeadersWhen: String => Boolean,
+  attributesFromResponse: Getter[Response[Unit], Map[String, AttributeValue]]
```
Comment (catostrophe) [inline modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala:23]:
```suggestion
  responseAttributesGetter: Getter[Response[_], Map[String, AttributeValue]]
```

File: modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala [hunk 4]
```diff
@@ -35,19 +39,23 @@ class SttpBackendTracer[F[_], G[_], +P, Ctx](
...
-            }
+          for {
+            _ <- childSpan.putAll(
+              SttpHeaders.requestFields(Headers(req.headers), dropHeadersWhen) ++ requestAttributes: _*
+            )
+            resp <- lower(ctxBackend.send(req))
+            _ <- childSpan.setStatus(SttpStatusMapping.statusToSpanStatus(resp.statusText, resp.code))
+            _ <- childSpan.putAll(SttpHeaders.responseFields(Headers(resp.headers), dropHeadersWhen): _*)
+            // attributesFromResponse could be expensive, so only call if the span is sampled
+            _ <- F.whenA(isSampled)(childSpan.putAll(attributesFromResponse.get(resp.copy(body = ()))))
```
Comment (catostrophe) [inline modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala:57]:
```suggestion
            _ <- F.whenA(isSampled)(childSpan.putAll(attributesFromResponse.get(resp)))
```

File: modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala [hunk 5]
```diff
@@ -35,19 +39,23 @@ class SttpBackendTracer[F[_], G[_], +P, Ctx](
...
-            .flatTap { resp =>
-              childSpan.setStatus(SttpStatusMapping.statusToSpanStatus(resp.statusText, resp.code))
-            }
+          for {
+            _ <- childSpan.putAll(
+              SttpHeaders.requestFields(Headers(req.headers), dropHeadersWhen) ++ requestAttributes: _*
+            )
+            resp <- lower(ctxBackend.send(req))
+            _ <- childSpan.setStatus(SttpStatusMapping.statusToSpanStatus(resp.statusText, resp.code))
+            _ <- childSpan.putAll(SttpHeaders.responseFields(Headers(resp.headers), dropHeadersWhen): _*)
```
Comment (catostrophe) [inline modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala:55]: I think these 2 calls could be one `putAll`

Comment (ybasket) [reply modules/sttp-client3/src/main/scala/io/janstenpickle/trace4cats/sttp/client3/SttpBackendTracer.scala]: Done.

**Review by catostrophe — APPROVED** [review]

**Comment by catostrophe** [comment]
Fine! I just renamed a few vals to make them look consistent.

**Pull request merged**